### PR TITLE
Port t-w monitor package to g-w.

### PR DESCRIPTION
**Pull request by walac**
Based on taskcluster-worker monitor package, we implement monitoring
support on generic-worker.

### CODE REVIEW

**Comment by jonasfj** [comment]
Note. I was thinking you would just import from tc-worker... But this is decent too... Since this doesn't use govendor...

File: mounts.go [hunk 1]
```diff
@@ -335,6 +335,11 @@ func (taskMount *TaskMount) Start() *CommandExecutionError {
 		// If the problem is internal (e.g. can't mount a writable cache) then
 		// this is handled by a panic.
 		if err != nil {
+			if monitor != nil {
+				incidentID := monitor.ReportError(err, "Error mounting directory")
+				return Failure(fmt.Errorf("%v, incident ID: %s", err, incidentID))
+			}
+
```
Comment (petemoore) [inline mounts.go]: I think we should only report errors for exceptions, not failures (this would be a task failure, and can be e.g. if an invalid url is specified for the preloaded content of the mount). There should be other places where we explicitly panic in the code.

File: main.go [hunk 2]
```diff
@@ -388,6 +404,9 @@ func loadConfig(filename string, queryUserData bool) (*Config, error) {
 		TasksDir:                       defaultTasksDir(),
 		WorkerGroup:                    "test-worker-group",
 		WorkerTypeMetadata:             map[string]interface{}{},
+		Project:                        "",
+		LogLevel:                       "debug",
+		Tags:                           "",
```
Comment (petemoore) [inline main.go]: Should tags be a string array, rather than a string?

Also, all these config variables need to be defined in the help text at the top of `main.go`. Note config is defined in two sections - mandatory and optional parameters - make sure to add to the correct section.

File: main.go [hunk 3]
```diff
@@ -388,6 +404,9 @@ func loadConfig(filename string, queryUserData bool) (*Config, error) {
 		TasksDir:                       defaultTasksDir(),
 		WorkerGroup:                    "test-worker-group",
 		WorkerTypeMetadata:             map[string]interface{}{},
+		Project:                        "",
```
Comment (petemoore) [inline main.go]: Since empty project is not allowed, maybe default can be something like `"test-project"`.

File: main.go [hunk 4]
```diff
@@ -315,7 +317,21 @@ func main() {
...
+				Certificate: config.Certificate,
+			}),
+			config.LogLevel,
+			nil,
+		)
+
+		exitCode := INTERNAL_ERROR
+		incidentID := monitor.CapturePanic(func() {
+			exitCode = RunWorker()
+		})
```
Comment (petemoore) [inline main.go]: I think I prefer capturing the panic inside `RunWorker` as before, since all of the tests use `RunWorker` as their entry point - thus, we'd lose the panic handling in tests if we do it here. The main method is very slim and well tested, so I do not think it is possible for a panic to occur there.

File: main.go [hunk 5]
```diff
@@ -464,14 +485,6 @@ func RunWorker() (exitCode ExitCode) {
...
 	}
 
-	defer func() {
-		if r := recover(); r != nil {
-			log.Print(string(debug.Stack()))
-			log.Print(" *********** PANIC occurred! *********** ")
-			log.Printf("%v", r)
-			exitCode = INTERNAL_ERROR
-		}
-	}()
```
Comment (petemoore) [inline main.go]: As mentioned above - all the tests call `RunWorker()` so let's keep the panic handling inside of it.

File: monitor.go [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	raven "github.com/getsentry/raven-go"
+	"github.com/pborman/uuid"
+	"github.com/taskcluster/statsum"
+	tcauth "github.com/taskcluster/taskcluster-client-go/auth"
+)
+
+// NewMonitor creates a new monitor
```
Comment (petemoore) [inline monitor.go]: I think this comment serves no purpose. Could you elaborate on what a monitor is, whether it can be called concurrently, how it should be used, etc?

File: monitor.go [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+		panic(fmt.Sprintf("Unsupported log-level: %s", logLevel))
+	}
+
+	// Convert tags to logrus.Fields
+	fields := make(logrus.Fields, len(tags))
+	for k, v := range tags {
+		fields[k] = v
+	}
+
+	// Declare monitor so we can reference it in OnError
```
Comment (petemoore) [inline monitor.go]: Note, you can achieve this too by declaring it as a return variable, i.e.

```
func NewMonitor(project string, auth *tcauth.Auth, logLevel string, tags map[string]string) (m *Monitor) {
```

That also simplifies the return statement to just `return`.

File: monitor.go [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+type sentry struct {
+	client     *raven.Client
+	m          sync.Mutex
+	project    string
+	expiration time.Time
+	auth       *tcauth.Auth
+}
+
+func (s *sentry) Client() (*raven.Client, error) {
```
Comment (petemoore) [inline monitor.go]: Here `sentry` is not exported, so `Client()` is not available outside the package, so maybe this should be `client()`? Note, the generic-worker is horribly inconsistent everywhere, but I just wanted to call it out as it seemed strange having an unexported type with an exported method (or is it satisfying an interface?)

File: monitor.go [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+			return nil, err
+		}
+		// Set expiration, so we remember to refresh
+		s.expiration = time.Time(res.Expires)
+	}
+
+	return s.client, nil
+}
+
+type Monitor struct {
```
Comment (petemoore) [inline monitor.go]: A comment here would be useful, also for the members (e.g. what is the prefix, why are tags map[string]string here, but in the json it is just a string?)

File: monitor.go [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+		m.Error("Failed to send error: ", err)
+		return
+	}
+
+	// Send packet
+	_, done := client.Capture(packet, tags)
+	<-done
+}
+
+func (m *Monitor) WithTags(tags map[string]string) *Monitor {
```
Comment (petemoore) [inline monitor.go]: Doc comment would be useful here.

File: monitor.go [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+func (m *Monitor) ReportWarning(err error, message ...interface{}) string {
+	incidentID := uuid.NewRandom()
+	m.Entry.WithField("incidentId", incidentID.String()).WithError(err).Warn(message...)
+	m.submitError(err, fmt.Sprint(message...), raven.WARNING, incidentID, 1)
+	return incidentID.String()
+}
+
+func (m *Monitor) submitError(err error, message string, level raven.Severity, incidentID uuid.UUID, skipFrames int) {
```
Comment (petemoore) [inline monitor.go]: Doc comment would be useful here.

File: monitor.go [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+
+type Monitor struct {
+	*statsum.Statsum
+	*logrus.Entry
+	*sentry
+	tags   map[string]string
+	prefix string
+}
+
+func (m *Monitor) Time(name string, fn func()) {
```
Comment (petemoore) [inline monitor.go]: Doc comment would be useful here (I know it is simple, but I've no idea what the statsum Time function does).

File: monitor_test.go [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+	require.True(t, incidentID == "")
+
+	incidentID = m.CapturePanic(func() {
+		callingSomethingBad()
+	})
+	require.True(t, incidentID != "")
+}
+
+func badThingHappens() {
+	panic("Oh, this is bad")
```
Comment (petemoore) [inline monitor_test.go]: love it!

**Comment by petemoore** [comment]
Also please request a second review when you've made the changes - and this time I promise to be much quicker reviewing - I'm really sorry the last one took so long to get around to (it arrived just in the middle of a complete overhaul/redesign of the worker). Thanks!

**Comment by petemoore** [comment]
P.P.S. Thanks for not importing taskcluster-worker as I think in this case, pulling it out was the cleaner and better solution. I think [Rob Pike would agree](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=9m28s). :-)

File: main.go (hunk 4 above)
Comment (walac) [inline main.go]: I don't remember exactly why, I remember I moved `CapturePani`c here because doing it inside `RunWorker` would make things more complicated.

File: main.go (hunk 3 above)
Comment (walac) [inline main.go]: This is to make sure whoever is using g-w remembers to set up a proper project name.

File: monitor.go (hunk 8 above)
Comment (walac) [inline monitor.go]: I adapted the code from tc-worker, so this could make sense there but not here.

File: mounts.go (hunk 1 above)
Comment (walac) [inline mounts.go]: iirc, I copied the behavior from tc-worker, which reports fatal errors.

File: main.go (hunk 3 above)
Comment (petemoore) [inline main.go]: What about for dev environments, where they don't want to hit production sentry etc?

File: mounts.go (hunk 1 above)
Comment (petemoore) [inline mounts.go]: This isn't a fatal error - it is just a task failure.

File: main.go (hunk 4 above)
Comment (petemoore) [inline main.go]: If a monitor instance isn't created during tests, is this handled gracefully?

The RunWorker method is designed not to panic, so I don't like it being able to panic now, which is a change in behaviour that could affect test results.

File: main.go (hunk 3 above)
Comment (petemoore) [inline main.go]: I prefer either:

a) not hitting statsum / sentry if a value isn't defined (which is probably what we want for dev environments)
b) using a default like `test-monitor` such that we have a defined bucket for assigning scopes (e.g. the role for running tests includes the scope to use `test-monitor` in the same way it grants the use of `test-provisioner` etc).

File: main.go (hunk 4 above)
Comment (walac) [inline main.go]: > If a monitor instance isn't created during tests, is this handled gracefully?
Do you mean monitoring calls?

>  The RunWorker method is designed not to panic, so I don't like it being able to panic now, which is a change in behavior that could affect test results.
Hrm, if a test expects `RunWorker` to panic, shouldn't it capture it? `Panic` is something orthogonal and unexpected by nature, so caller either shouldn't expect any function to panic or expect any function to...

File: main.go (hunk 2 above)
Comment (walac) [inline main.go]: It is actually not currently used, removed.

File: monitor.go (hunk 6 above)
Comment (walac) [inline monitor.go]: Added to the `Monitor` type

File: artifacts.go [hunk 14]
```diff
@@ -121,7 +121,8 @@ func (artifact *S3Artifact) CreateTempFileForPUTBody() string {
 	baseName := filepath.Base(rawContentFile)
 	tmpFile, err := ioutil.TempFile("", baseName)
 	if err != nil {
-		log.Fatal(err)
+		incidentID := monitor.ReportError(err, "Could not create temporary file")
+		log.Fatalf("%v, incident ID: %s", err, incidentID)
```
Comment (petemoore) [inline artifacts.go]: I've only just realised this myself, but I have a few places in the code where I am calling e.g. `log.Fatal` which should really just panic, so that the worker shutdown etc is handled gracefully (since we catch the panics and handle them). This PR is about something else, so I won't ask you to fix all my broken code, but I'll do that myself once this lands. In other words, where there are currently `log.Fatal[f]` function calls that execute _in the main go routine_, I'll change them to `panic` statements instead.

File: main.go [hunk 15]
```diff
@@ -456,14 +476,6 @@ func loadConfig(filename string, queryUserData bool) (*Config, error) {
 }
 
 func RunWorker() (exitCode ExitCode) {
-	log.Printf("Detected %s platform", runtime.GOOS)
```
Comment (petemoore) [inline main.go]: Why has this been removed?

File: monitor.go [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+// A Monitor is responsible for collecting logs, stats and error messages.
+//
+// A monitor is a context aware object for monitoring. That is to say that a
+// Monitor is used to record metrics, write logs and report errors. When doing
+// so the Monitor object adds meta-data to the metrics, logs and errors. The
+// meta-data added is context dependent tags and prefix. These help identify
+// where a log message, metric or error originates from.
+//
+// By encapsulating the context meta-data inside the Monitor object, an
+// implementor gets a Monitor rarely needs to add tags or prefix. For example
```
Comment (petemoore) [inline monitor.go]: This first sentence doesn't make sense.

File: monitor.go [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+// a monitor will always be prefixed by plugin name before being passed to a
+// plugin, hence, it is easy trace any log message, metric or error report to
+// the plugin that it was created in.
```
Comment (petemoore) [inline monitor.go]: Can you reword "plugin" as "feature"? ("plugin" is the name we use in taskcluster-worker, but in generic-worker we called the equivalent thing a "feature")

**Comment by petemoore** [comment]
> Note. I was thinking you would just import from tc-worker... But this is decent too... Since this doesn't use govendor...

@jonasfj I think it would be a shame to pull in the entire taskcluster-worker repo just for this. See also [this](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=9m28s).

**Comment by petemoore** [comment]
@walac Also, can you clarify the reason that we have the panic handling in the `main` method rather than just keeping it in the `RunWorker` function (and why to add the doRunWorker function)? Thanks!

**Comment by petemoore** [comment]
@walac Also looks like there are merge conflicts that need resolving... Thanks!

**Comment by walac** [comment]
> @walac Also, can you clarify the reason that we have the panic handling in the `main` method rather than just keeping it in the `RunWorker` function (and why to add the doRunWorker function)? 
When running tests, we capture panic through `defer`/`recover`, when running the worker, we need to capture through monitoring. This preserves the previous behavior of `RunWorker`. `doRunWorker` is just a helper to make the code simpler.

File: main.go (hunk 4 above)
Comment (petemoore) [inline main.go]: Tests don't expect RunWorker to panic, since RunWorker doesn't panic. However, they check the exit code that RunWorker returns, and if a panic occurs, this will be captured in the exitCode. They do check this. So really it is about the RunWorker method being designed to return a code based on what happened (idle timeout, worker requests a reboot, worker crashed, job was successful, etc) - and tests check that. Does that make sense?

**Comment by petemoore** [comment]
The monitor initialisation and panic capture should occur in `RunWorker`, since currently if you call RunWorker (e.g. from a test) we will have a nil pointer exception for any call to a monitor's methods. We don't need doRunWorker - we just need to move the monitor initialisation logic into RunWorker. The global config will always provide the values to use, and in RunWorker, you will always need the values set, so there is no reason to initialise the monitor outside of the RunWorker function, and no reason to create the doRunWorker function.

Additionally, no logging calls have been adapted to use the new Monitor, so none of the existing logging will go through the monitor. If we are not using the logger, let's get rid of it. If we intend to use it, please replace the old logger calls - otherwise we have two different loggers in one codebase.

In the PR I see only a single call on the Monitor object, which is to report a crash to sentry. The other features are implemented, but then not used. I don't see an advantage to port all this code from taskcluster-worker, and then not use it. Since we already have a code block handling exceptions and crashes (panics), it would make more sense to directly report to sentry here, and not introduce all the new abstractions that were designed for a different worker.

In taskcluster-worker there is a concept of having a component hierarchy, which can be used to identify a component when logging information. In generic-worker, there is no such component hierarchy, so the code around this is probably unneeded (since it is not called).

My concern with this PR is it brings over unnecessary complexity from taskcluster-worker, that isn't suited to generic-worker, it provides methods which are never called, and introduces more dependencies. Since it only adds reporting crashes to sentry, maybe it would be best just to add that in a couple of lines of code instead in the current code block that currently logs the panic to papertrail and to the task log.

I don't see an advantage of changing the logging mechanism. The code for reporting to statsum is something that could be useful, but I don't see a good reason for it to be embedded in the monitor object - as far as I see, uploading crash reports is a separate concern to recording activity metrics. The two things use separate systems, there is no code commonality, I think having a Monitor object that handles both these unrelated things makes the design more brittle and introduces false dependencies. Since these are separate concerns, I'd propose we handle them via different interfaces, which can be maintained independently.

In other words, what I'm really asking:

1) let's leave logging as it is, as it does what we want already
2) let's configure a global sentry client, and then use it to report some real (useful) data (i.e. not just create it)
3) let's adapt the existing top-level panic handling routine to also send crash reports to sentry (in addition to what it is already doing)

**Comment by jonasfj** [comment]
@petemoore,
If you don't want the complexity monitor brings with it... just use the
statsum client for golang and sentry client directly.
Neither one is very hard to use... and monitor demonstrates how to do it...

--
Regards Jonas Finnemann Jensen.

2017-06-09 6:26 GMT-07:00 Pete Moore <notifications@github.com>:

File: main.go [hunk 18]
```diff
@@ -24,6 +24,7 @@ import (
 	"github.com/taskcluster/taskcluster-base-go/scopes"
 	tcclient "github.com/taskcluster/taskcluster-client-go"
 	"github.com/taskcluster/taskcluster-client-go/auth"
+	tcauth "github.com/taskcluster/taskcluster-client-go/auth"
```
Comment (petemoore) [inline main.go]: we might not need this

**Comment by petemoore** [comment]
Let's go with this, and delete unused code in a future PR. Thanks!
